src/deim.jl

```diff
@@ -0,0 +1,167 @@
+- `nonlinear_projection_matrix::AbstractMatrix`: the projection matrix ``\\underset{n\\times m}U`` for the nonlinear functions ``\\mathbf F``.
+
+# Return
+- `reduced_rhss`: the right hand side of ROM.
+- `linear_projection_eqs`: the linear projection mapping ``\\mathbf y=V\\hat{\\mathbf y}``.
+"""
+function deim(full_vars::AbstractVector, linear_coeffs::AbstractMatrix,
+              constant_part::AbstractVector, nonlinear_part::AbstractVector,
+              reduced_vars::AbstractVector, linear_projection_matrix::AbstractMatrix,
+              nonlinear_projection_matrix::AbstractMatrix)
+    # rename variables for convenience
+    y = full_vars
+    A = linear_coeffs
+    g = constant_part
+    F = nonlinear_part
+    ŷ = reduced_vars
+    V = linear_projection_matrix
+    U = nonlinear_projection_matrix
+
+    # original vars to reduced vars
+    linear_projection_eqs = Symbolics.scalarize(y .~ V * ŷ)
+    linear_projection_dict = Dict(eq.lhs => eq.rhs for eq in linear_projection_eqs)
+
+    indices = deim_interpolation_indices(U) # DEIM interpolation indices
+    # the DEIM projector (not DEIM basis) satisfies
+    # F(original_vars) ≈ projector * F(pod_basis * reduced_vars)[indices]
+    projector = ((@view U[indices, :])' \ (U' * V))'
+    temp = substitute.(F[indices], (linear_projection_dict,))
+    F̂ = projector * temp # DEIM approximation for nonlinear func F
+
+    Â = V' * A * V
+    ĝ = V' * g
+    reduced_rhss = Â * ŷ + ĝ + F̂
+    reduced_rhss, linear_projection_eqs
+end
+"""
+$(TYPEDSIGNATURES)
+
+Reduce a `ModelingToolkit.ODESystem` using the Proper Orthogonal Decomposition (POD) with
+the Discrete Empirical Interpolation Method (DEIM).
+
+`snapshot` should be a matrix with the data of each time instance as a column.
+
+The LHS of equations in `sys` are all assumed to be 1st order derivatives. Use
+`ModelingToolkit.ode_order_lowering` to transform higher order ODEs before applying DEIM.
+
+`sys` is assumed to have no internal systems. End users are encouraged to call
+`ModelingToolkit.structural_simplify` beforehand.
+
+The POD basis used for DEIM interpolation is obtained from the snapshot matrix of the
+nonlinear terms, which is computed by executing the runtime-generated function for
+nonlinear expressions.
+"""
+function deim(sys::ODESystem, snapshot::AbstractMatrix, pod_dim::Integer;
+              deim_dim::Integer = pod_dim,
+              name::Symbol = Symbol(nameof(sys), :_deim))::ODESystem
+    @set! sys.name = name
+
+    # handle ODESystem.substitutions
+    # https://github.com/SciML/ModelingToolkit.jl/issues/1754
+    sys = tearing_substitution(sys)
+
+    iv = ModelingToolkit.get_iv(sys) # the single independent variable
+    D = Differential(iv)
+    dvs = ModelingToolkit.get_states(sys) # dependent variables
+
+    pod_reducer = POD(snapshot, pod_dim)
+    reduce!(pod_reducer, TSVD())
+    V = pod_reducer.rbasis # POD basis
+
+    var_name = gensym(:ŷ)
```

Comment:
a pure gensym will recompile each time. Let's get an issue open about replacing this with a hashing version. no need to solve this in this PR though.

Reply:
Did this get an issue?